execution-engine/src/fs.rs

```diff
@@ -0,0 +1,1057 @@
+    pub(crate) fn fd_read(&mut self, fd: Fd, len: usize) -> FileSystemError<Vec<u8>> {
+        self.check_right(&fd, Rights::FD_READ)?;
+        let offset = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?.offset;
+
+        let rst = self.fd_pread(fd, len, offset)?;
+        self.fd_seek(fd, rst.len() as i64, Whence::Current)?;
+        Ok(rst)
+    }
+
+    /// The stub implementation of `fd_readdir`. Return unsupported error `NoSys`.
+    #[inline]
+    pub(crate) fn fd_readdir(&mut self, fd: Fd, _cookie: DirCookie) -> FileSystemError<Vec<DirEnt>> {
+        self.check_right(&fd, Rights::FD_READDIR)?;
+        Err(ErrNo::NoSys)
+    }
+
+    /// Atomically renumbers the `old_fd` to the `new_fd`.  Note that as
+    /// execution engine is single-threaded this is atomic from the WASM program's
+    /// point of view.
+    pub(crate) fn fd_renumber(&mut self, old_fd: Fd, new_fd: Fd) -> FileSystemError<()> {
+        let entry = self.fd_table.get(&old_fd).ok_or(ErrNo::BadF)?.clone();
+        if self.fd_table.get(&new_fd).is_none() {
+            self.fd_table.insert(new_fd, entry);
+            self.fd_table.remove(&old_fd);
+            Ok(())
+        } else {
+            Err(ErrNo::BadF)
+        }
+    }
+
+    /// Change the offset of Fd `fd`.
+    pub(crate) fn fd_seek(
+        &mut self,
+        fd: Fd,
+        delta: FileDelta,
+        whence: Whence,
+    ) -> FileSystemError<FileSize> {
+        self.check_right(&fd, Rights::FD_SEEK)?;
+        let FileTableEntry { inode, offset, .. } = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?;
+        let file_size = self
+            .inode_table
+            .get(inode)
+            .ok_or(ErrNo::NoEnt)?
+            .file_stat
+            .file_size;
+
+        let new_base_offset = match whence {
+            Whence::Current => *offset,
+            Whence::End => file_size,
+            Whence::Start => 0,
+        };
+
+        // NOTE: Ensure the computation does not overflow.
+        let new_offset: FileSize = if delta >= 0 {
+            // It is safe to convert a positive i64 to u64.
+            let t_offset = new_base_offset + (delta.abs() as u64);
+            // Offset is allowed to equal to file size
+            if t_offset > file_size {
+                return Err(ErrNo::SPipe);
+            }
+            t_offset
+        } else {
+            // It is safe to convert a positive i64 to u64.
+            if (delta.abs() as u64) > new_base_offset {
+                return Err(ErrNo::SPipe);
+            }
+            new_base_offset - (delta.abs() as u64)
+        };
+
+        // Update the offset
+        self.fd_table.get_mut(&fd).ok_or(ErrNo::BadF)?.offset = new_offset;
+        Ok(new_offset)
+    }
+
+    /// The stub implementation of `fd_sync`. Return unsupported error `NoSys`.
+    #[inline]
+    pub(crate) fn fd_sync(&mut self, fd: Fd) -> FileSystemError<()> {
+        self.check_right(&fd, Rights::FD_SYNC)?;
+        Err(ErrNo::NoSys)
```

Comment:
Shouldn't we provide `fd_sync`? It is just a noop.

Reply:
Done